api/simqueue/db.py

```diff
@@ -173,31 +298,271 @@ async def query_jobs(
+    #compiled_query=str(results.compile(compile_kwargs={"literal_binds": True}))
+    #return(results)
+    #return [dict(result) for result in results]
+    if fields:
+        return [dict(result) for result in results]
     return [transform_fields(await follow_relationships(dict(result))) for result in results]
 
 
 async def get_job(job_id: int):
     query = jobs.select().where(jobs.c.id == job_id)
     result = await database.fetch_one(query)
+    if result is not None:
+        intermediate_result =dict(result)
+        return transform_fields(await follow_relationships(intermediate_result))
+    else:
+        raise HTTPException(status_code=status_codes.HTTP_404_NOT_FOUND,
+        detail=f"Either there is no job with id {job_id}, or you do not have access to it"
+    )
+
+async def post_job(user_id:str,job:SubmittedJob):
+
+    ins = jobs.insert().values( code=job.code,command=job.command,collab_id=job.collab_id,user_id=user_id,status="submitted",hardware_platform=job.hardware_platform,hardware_config=job.hardware_config,timestamp_submission=now_in_utc())
+    job_id = await database.execute(ins)
+    
+    query = jobs.select().where(jobs.c.id == job_id)
+    result = await database.fetch_one(query)
+    if job.input_data is not None:
+        await create_job_input_data_item(job,job_id)
+    if job.tags is not None:
+        list_tags =[item["tag_id"] for item in job.tags or [] ]
+        await post_tags(job_id, list_tags)
+    return transform_fields(await follow_relationships(dict(result)))
+
+async def put_job(job_id:int,user_id:str,jobPatch:JobPatch):
+    query = jobs.select().where(jobs.c.id == job_id)
+    job = await database.fetch_one(query)
+    ins = jobs.update().where(jobs.c.id == job_id).values(provenance=jobPatch.provenance,resource_usage=jobPatch.resource_usage)
+    #ins = jobs.update().where(jobs.c.id == job_id).values(id=job_id,code=job.code,command=job.command,collab_id=job.collab_id,user_id=user_id,status=jobPatch.status,hardware_platform=job.hardware_platform,hardware_config=job.hardware_config,timestamp_submission=job.timestamp_submission,timestamp_completion=job.timestamp_completion,provenance=jobPatch.provenance,resource_usage=jobPatch.resource_usage,log=jobPatch.log,output_data=jobPatch.output_data)
+    #ins = jobs.insert().values( code=job.code,command=job.command,collab_id=job.collab_id,user_id=user_id,status="submitted",hardware_platform=job.hardware_platform,hardware_config=job.hardware_config,timestamp_submission=now_in_utc())
+    await database.execute(ins)
+    
+    query = jobs.select().where(jobs.c.id == job_id)
+    result = await database.fetch_one(query)
+    if jobPatch.output_data is not None:
+        await update_job_output_data_item(result,jobPatch)
+    if jobPatch.log is not None:
+        await update_log(result,jobPatch)
     return transform_fields(await follow_relationships(dict(result)))
 
+async def delete_job(job_id: int):
+    query = jobs.select().where(jobs.c.id == job_id)
```

Comment:
I think you could avoid the double query here by rewriting `delete_dataitems()` to take `job_id` as its argument rather than `job`